hrc_speech_prediction/models.py

```diff
@@ -90,12 +150,183 @@ def fit(self, X_context, X_speech, labels):
+            X_context, X_speech)
+        l_probas = (
+            np.power(probas, self.beta) * self._p[None, :] / self._q[None, :])
         return normalize(l_probas, axis=1)
+
+
+class CombinedModel(object):
+    def __init__(self,
+                 vectorizer,
+                 model_generator,
+                 actions,
+                 speech_eps=0.15,
+                 context_eps=0.15):
+
+        self._speech_eps = speech_eps
+        self._context_eps = context_eps
+
+        self.actions = actions
+
+        self.model_generator = model_generator
+        self._vectorizer = vectorizer
+
+        self.context_model = context_model.ContextTreeModel(
+            self.actions, eps=context_eps)
+        self.speech_model = None
+
+        self._X_context = np.ones((1, len(self.actions)), dtype='bool')
+
+    def fit(self, ctxt, speech, actions):
+        self.context_model.fit(ctxt, actions)
+        self.speech_model = self.model_generator(
+            self.actions, features="speech").fit(
+                self._X_context, speech, actions, sample_weight=None)
+
+    def partial_fit(self, ctxt, speech, action):
+
+        if isinstance(speech, str):
+            x_u = self._vectorizer.transform([speech])
+        else:
+            x_u = speech  # Then the input is an numpy array already
+
+        self.context_model.fit(ctxt, action)
+
+        if self.speech_model:
+            self.speech_model.partial_fit(self._X_context, x_u, action,
+                                          self.actions)
+        else:
+            self.speech_model = self.model_generator(
+                self.actions, features="speech").partial_fit(
+                    self._X_context, x_u, action, classes=self.actions)
+
+    def predict(self,
+                cntxt,
+                utter=None,
+                model="both",
+                exclude=None,
+                plot=False,
+                return_probs=False):
+
+        if model == "context" is not None:
+            probs = self.get_context_probs(cntxt)
+        elif model == "speech" and utter is not None:
+            probs = self.get_speech_probs(utter)
+        elif model == "both" and utter is not None:
+            context_probs = self.get_context_probs(cntxt)
+            speech_probs = self.get_speech_probs(utter)
+        else:
+            raise "Error, bad inputs to predict()!"
+
+        probs = np.multiply(context_probs, speech_probs)
+
+        action = self.get_probable_action(probs, exclude)
+        # self.curr = self.curr._get_next_node(action)
+
+        if plot and model == "both":
+            all_probs = np.vstack((speech_probs, context_probs, probs))
+            names = ["Speech", "Context", "Combined"]
+
+            plt.gcf().clear()
+            plots.plot_predict_proba(
+                all_probs, self.actions, utter, model_names=names)
+            plt.show(block=False)
+        if return_probs and model == "both":
```

Comment:
Here it is annoying to change the return signature of the method for something (the nature of the model) that does not depend only on its argument. For example if you write a code that just wants the predicted action and does not care about the nature of the model (and hence should work regardless), you cannot use pattern matching to write: `action, _ = model.predict(...)` because the method may return two or for items and the former would only work for two.

Here, you could return either `action, (speech_probs, context_probs, probs)` or `action, probs` which would be more consistant. Another way to do it would be returning the stacked vector of probabilities, that way only the dimensions of the vector would differ.